### PR TITLE
Histogram Rollout - Algorithms 93-97

### LINKED ISSUE #17737: HistogramData Rollout: Algorithms 93-97
HistogramData rollout: algorithms 93-97

This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:

```
Framework/CurveFitting/src/IMWDomainCreator.cpp
Framework/CurveFitting/src/SeqDomainSpectrumCreator.cpp
Framework/DataHandling/src/AsciiPointBase.cpp
Framework/DataHandling/src/CreateSimulationWorkspace.cpp
Framework/DataHandling/src/GroupDetectors2.cpp
```

### PR DESCRIPTION
**Pull request by DavidFair**
HistogramData rollout: algorithms 93-97
## Description of work.

This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

```
Framework/CurveFitting/src/IMWDomainCreator.cpp
Framework/CurveFitting/src/SeqDomainSpectrumCreator.cpp
Framework/DataHandling/src/AsciiPointBase.cpp
Framework/DataHandling/src/CreateSimulationWorkspace.cpp
Framework/DataHandling/src/GroupDetectors2.cpp
```

Performance tests were added for:

```
CreateSimulationWorkspace
GroupDetectors2
```

The following algorithms have improved performance:
None
## To test.

Code review.

Fixes #17737 .
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

### CODE REVIEW

File: Framework/DataHandling/src/AsciiPointBase.cpp [hunk 1]
```diff
@@ -59,7 +59,7 @@ void AsciiPointBase::exec() {
  */
 std::vector<double> AsciiPointBase::header(std::ofstream &file) {
   auto title = '#' + m_ws->getTitle();
-  const std::vector<double> &xTemp = m_ws->readX(0);
+  const auto &xTemp = m_ws->x(0);
   m_xlength = xTemp.size() - 1;
   std::vector<double> XData(m_xlength, 0);
   for (size_t i = 0; i < m_xlength; ++i) {
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/AsciiPointBase.cpp]: `auto XData = m_ws->points(0)` should do the job, but the type is `Histogram::Points`. Have you considered changing the return type of this function?

File: Framework/DataHandling/src/CreateSimulationWorkspace.cpp [hunk 2]
```diff
@@ -187,11 +187,9 @@ void CreateSimulationWorkspace::createOutputWorkspace() {
   PARALLEL_FOR1(m_outputWS)
   for (int64_t i = 0; i < static_cast<int64_t>(nhistograms); ++i) {
     m_outputWS->setBinEdges(i, binBoundaries);
-    MantidVec &yOut = m_outputWS->dataY(i);
-    for (size_t j = 0; j < ylength; ++j) {
-      yOut[j] = 1.0; // Set everything to a value so that you can visualize the
-                     // output sensibly
-    }
+    Mantid::HistogramData::HistogramY yOut(ylength, 1.0);
+    m_outputWS->mutableY(i) = yOut;
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/CreateSimulationWorkspace.cpp]: `m_outputWS->mutableY(i) = 1.0;`

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 3]
```diff
@@ -240,8 +240,8 @@ void GroupDetectors2::execEvent() {
   }
 
   // Set all X bins on the output
-  auto XValues = HistogramData::BinEdges(inputWS->readX(0));
-  outputWS->setAllX(XValues);
+  auto binEdges = inputWS->binEdges(0);
+  outputWS->setAllX(binEdges);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: Even shorter: `outputWS->setAllX(inputWS->binEdges(0));`, no need for the temporary.

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 4]
```diff
@@ -936,23 +936,23 @@ size_t GroupDetectors2::formGroups(API::MatrixWorkspace_const_sptr inputWS,
...
       // detectors to add to firstSpecNum
       const auto &fromSpectrum = inputWS->getSpectrum(originalWI);
 
       // Add up all the Y spectra and store the result in the first one
-      auto fEit = outSpec.dataE().begin();
-      auto Yit = fromSpectrum.dataY().cbegin();
-      auto Eit = fromSpectrum.dataE().cbegin();
+      auto fEit = outSpec.mutableE().begin();
+      auto Yit = fromSpectrum.y().cbegin();
+      auto Eit = fromSpectrum.e().cbegin();
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: There is `operator+` for histograms (you find it in `MantidHistogramData/HistogramMath.h`). The code here assumes that the bin boundaries for all spectra are the same, whereas `operator+` actually checks that. Could you try using `+` here and see if any tests fail?

**Review by SimonHeybrock — COMMENTED** [review]
Looks good apart from small comments.

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 5]
```diff
@@ -938,40 +938,21 @@ size_t GroupDetectors2::formGroups(API::MatrixWorkspace_const_sptr inputWS,
...
-    beh->mutableX(outIndex)[0] = 0.0;
-    beh->mutableE(outIndex)[0] = 0.0;
+
     for (auto originalWI : it->second) {
       // detectors to add to firstSpecNum
-      const auto &fromSpectrum = inputWS->getSpectrum(originalWI);
+      const auto &inputSpectrum = inputWS->getSpectrum(originalWI);
 
-      // Add up all the Y spectra and store the result in the first one
+      outSpec.setHistogram(outSpec.histogram() += inputSpectrum.histogram());
```
Comment (DavidFair) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: Doh - A case of premature optimisation as in my head I was eliminating a variable.

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 6]
```diff
@@ -938,40 +938,21 @@ size_t GroupDetectors2::formGroups(API::MatrixWorkspace_const_sptr inputWS,
...
-
-      // detectors to add to the output spectrum
-      outSpec.addDetectorIDs(fromSpectrum.getDetectorIDs());
-      try {
-        Geometry::IDetector_const_sptr det = inputWS->getDetector(originalWI);
-        if (!det->isMasked())
-          ++nonMaskedSpectra;
-      } catch (Exception::NotFoundError &) {
-        // If a detector cannot be found, it cannot be masked
+      if (!isMaskedDetector(*inputWS->getDetector(originalWI))) {
```
Comment (DavidFair) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: Does `SpectrumInfo` throw as I might just change the call in the wrapper if it still does?

Comment (SimonHeybrock) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: With `SpectrumInfo` the best thing to do is to get rid of the exception-based control flow (exeptions are not for control flow!), so you could write something like:

```cpp
if (spectrumInfo.hasDetectors(i)) {
  if (spectrumInfo.isMasked(i)
    // bla
} else {
  // handle failure
}
```

**Comment by SimonHeybrock** [comment]
Looks good now. I wonder if the tests will pass (they should, unless some client code is doing dubious things with `GroupDetectors`).

**Comment by SimonHeybrock** [comment]
:shipit: !

**Pull request merged**